IBC.md

```diff
@@ -305,10 +308,101 @@ pub struct IbcPacket {
 }
 ```
 
-TODO: explain how to handle/parse errors (As part of
-https://github.com/CosmWasm/cosmwasm/issues/762)
+##### Acknowledgement Processing
+
+A major issue that is unique to `ibc_packet_receive` is that it is expected to
+often reject an incoming packet, but it cannot abort the transaction. We
+actually expect all state changes from the contract (as well as dispatched
+messages) to be reverted when the packet is rejected, but the transaction to
+properly commit an acknowledgement with encoded error. In other words, this "IBC
+Handler" will error and revert, but the "IBC Router" must succeed and commit an
+acknowledgement message (that can be parsed by the sending chain as an error).
+
+The atomicity issue was first
+[analyzed in the Cosmos SDK implementation](https://github.com/cosmos/ibc-go/issues/68)
+and refined into
+[changing semantics of the OnRecvPacket SDK method](https://github.com/cosmos/ibc-go/issues/91),
+which was
+[implemented in April 2021](https://github.com/cosmos/ibc-go/pull/107), likely
+to be released with Cosmos SDK 0.43 or 0.44. Since we want the best,
+future-proof interface for contracts, we will use an approach inspired by that
+work, and add an adapter in `wasmd` until we can upgrade to a Cosmos SDK version
+that implements this.
+
+After quite some
+[discussion on how to encode the errors](https://github.com/CosmWasm/cosmwasm/issues/762),
+we identified a number of issues implementing this idea with the CosmWasm model:
+
+1. We only return `Ok(Response)` or `Err(String)`. There is no way to return an
+   arbitrary binary encoded message in the error variant.
+2. If we return `messages` to be dispatched, they may fail with an error, which
+   is returned directly to the x/wasm handler. Obviously an error from the bank
+   handler will not be in the proper IBC Acknowledgement format.
```

Comment:
Is this paragraph about messages or sub-messages? Because for messages the transaction is terminated and reverted, such that no acknowledgement can be created, no matter if the format is known.

Reply:
> Because for messages the transaction is terminated and reverted, such that no acknowledgement can be created, no matter if the format is known.

Oh, this will change for this one entry point. That is the crux of this proposal. An error from a message will be handled like an error from `ibc_receive_packet`, which means it will be sent to the encoder.

Please let me know how I can express this clearer.

Reply:
I would not do this. This changes fundamental principles of how messages work. What's preventing `ibc_receive_packet` implementors to use sub messages instead of messages?